### PR TITLE
i#5538 skip: Fix skip-1-instr and skip-0-instr cases

### PR DESCRIPTION
**Pull request by derekbruening**
Fixes a boundary case of skipping 1 instruction when the scheduler has already read an instruction record but not yet passed it to the user.

Fixes a boundary case of back-to-back regions of interest.

Adds test cases.

Issue: #5538

### CODE REVIEW

File: clients/drcachesim/tests/scheduler_unit_tests.cpp [hunk 1]
```diff
@@ -470,17 +474,92 @@ test_regions_bare()
...
+// Tests regions without timestamps with an instr at the very front of the trace.
+static void
+test_regions_bare_no_marker()
+{
+    std::cerr << "\n----------------\nTesting bare regions with no marker\n";
+    std::vector<trace_entry_t> memrefs = {
+        /* clang-format off */
+        make_thread(1),
+        make_pid(1),
+        make_instr(1),
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scheduler_unit_tests.cpp:513]: Maybe a comment explicitly saying that this scenario is not possible in a real trace (the first entry being an instr not a marker).

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/scheduler/scheduler.h [hunk 2]
```diff
@@ -200,9 +200,10 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
          * though the input were constructed by concatenating these ranges together.  A
          * #TRACE_MARKER_TYPE_WINDOW_ID marker is inserted between
          * ranges (with a value equal to the range ordinal) to notify the client of the
-         * discontinuity (but not before the first range), with a
-         * #dynamorio::drmemtrace::TRACE_TYPE_THREAD_EXIT record inserted after the final
-         * range.  These ranges must be non-overlapping and in increasing order.
+         * discontinuity (but not before the first range nor between back-to-back regions
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:203]: My reasoning is that the marker is there to warn about a discontinuity in the instruction flow: a PC gap.  Back-to-back regions have no gap.   I thought the simulator normally would not really be aware of the regions and should treat their union as though it were one trace.  We can bring this up offline to see if there are use cases where the simulator would treat the regions differently.

File: clients/drcachesim/tests/scheduler_unit_tests.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tests/scheduler_unit_tests.cpp:513]: I added:
        // This would not happen in a real trace, only in tests.  But it does
        // match a dynamic skip from the middle when an instruction has already
        // been read but not yet passed to the output stream.

Skipping from the middle will have its own tests when added but this does give some confidence corner cases there may already be supported.

File: clients/drcachesim/scheduler/scheduler.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:203]: There is no real downside to adding the marker so I'm going ahead and adding it.
